### DISCUSSION TITLE
Enabling experimentalReplAwait in repl, and bindings

### DISCUSSION THREAD

**Discussion by tslocke** [TypeStrong/ts-node#1860] [Q&A]
If I start up a repl just with `ts-node`, I get top-level await support by default.

I'm trying to enable the same in a custom repl script but can't figure it out. Any advice? 

Script:

```js
import * as tsnode from 'ts-node'

const tsrepl = tsnode.createRepl()
const service = tsnode.register({
  ...tsrepl.evalAwarePartialHost,
  transpileOnly: true,
})
tsrepl.setService(service)

tsrepl.start()
```

Also is there a way to create bindings in the repl? Like node's `repl.context.foo = 123`

Thanks!

**Comment by cspotcode** [comment]
Hmm, does our API expose these things?  I think the experimental top level
await is an option somewhere.  The tricky thing about top level await, is
that it makes all execution possibly async.  Last I checked, node's own
REPL API does not expose this functionality.

Just a couple guesses to get you started.  If you feel like any of this
should become a feature request, and you feel confident describing it
clearly enough for someone to implement, then feel free to open an issue.

**Reply by tslocke** [reply]
Thanks for the advice. My use case is that I just want the functionality of the normal ts-node repl, but I want to run some initialising code first, specific to my project (e.g. import a bunch of stuff).

So given that top-level await seems to work OK in the default ts-node repl, I figure it must be doable. Just can't figure out where that option is.

I don't mind using private APIs - it's really just a throwaway convenience thing. I did try to browse the source to see how the regular repl is started up, but things got complex pretty quickly with all the options etc.

**Reply by cspotcode** [reply]
This is the option for top-level await:
https://typestrong.org/ts-node/docs/options#noexperimentalreplawait
https://typestrong.org/ts-node/api/interfaces/CreateOptions.html#experimentalReplAwait

**Reply by tslocke** [reply]
Thank you. I did find that in the docs but try as I might I couldn't get it working.

I just got to the bottom of it. `ts-node --version` in my terminal was telling me 10.9.1 but the version in node_modules was old. It's working great now. Sorry for the noise - I could have figured this out without posting had I noticed that.

For bindings, the script can just assign to `globalThis`.

Here's the final script and alias, in case anyone else is interested.

project-dir/.repl.ts
```ts
import * as tsnode from 'ts-node'

// Stuff I don't want to do every time I start a repl:
import {doSomething, initSomething} from '@/my-useful-module'
initSomething()
Object.assign(globalThis, {doSomething})

// Start the repl
const repl = tsnode.createRepl()
const service = tsnode.register({
  ...repl.evalAwarePartialHost,
  transpileOnly: true,
  experimentalReplAwait: true
})
repl.setService(service)
repl.start()
```

Alias
```
alias tsrepl='NODE_REPL=yes ts-node -r tsconfig-paths/register --transpileOnly ./.repl.ts'
```

Bliss!
